fireshape/control.py

```diff
@@ -657,6 +662,693 @@ def load(self, vec, filename="control.dat"):
+            M = np.hstack((M0, M1))
+            n = M.shape[0]
+            T[:n, :n] = M @ T[:n, :n]
+
+        n = 2**j0 + d - 1 - 2 * bc
+        T_split = [T[:, :n]]
+        n_split = [n]
+        offset = n
+        for j in range(j0, J):
+            n = 2**j
+            T_split.append(T[:, offset:offset+n])
+            n_split.append(n)
+            offset += n
+        self.n_split.append(n_split)
+        return T_split
+
+    def construct_1d_interpolation_matrices(self, V):
+        interp_1d = []
+
+        # this code is correct but can be made more beautiful
+        # by replacing x_fct with self.id
+        x_fct = fd.SpatialCoordinate(self.mesh_r)  # used for x_int
+        # compute self.M, x_int will be overwritten below
+        x_int = fd.interpolate(x_fct[0], V.sub(0))
+        self.M = x_int.vector().size()
+
+        comm = self.comm
+
+        u, v = fd.TrialFunction(V.sub(0)), fd.TestFunction(V.sub(0))
+        mass_temp = fd.assemble(u * v * fd.dx)
+        self.lg_map_fe = mass_temp.petscmat.getLGMap()[0]
+
+        for dim in range(self.dim):
+            order = self.orders[dim]
+            level = self.levels[dim]
+            T = self.mat[dim]
+            bc = self.boundary_regularities[dim]
+            knots = self.knots[dim]
+            m = 2**level + order - 1
+
+            interp_sub = []
+            for T_sub in T:
+                n = T_sub.shape[1]
+
+                # owned part of global problem
+                local_n = n // comm.size + int(comm.rank < (n % comm.size))
+                I = PETSc.Mat().create(comm=self.comm)
+                I.setType(PETSc.Mat.Type.AIJ)
+                lsize = x_int.vector().local_size()
+                gsize = x_int.vector().size()
+                I.setSizes(((lsize, gsize), (local_n, n)))
+
+                I.setUp()
+                x_int = fd.interpolate(x_fct[dim], V.sub(0))
+                x = x_int.vector().get_local()
+                for idx in range(n):
+                    coeffs = np.zeros(m, dtype=float)
+
+                    # impose boundary regularity
+                    coeffs[bc:m-bc] = T_sub[:, idx]
+                    degree = order - 1  # splev uses degree, not order
+                    tck = (knots, coeffs, degree)
+
+                    values = splev(x, tck, der=0, ext=1)
+                    rows = np.where(values != 0)[0].astype(np.int32)
+                    values = values[rows]
+                    rows_is = PETSc.IS().createGeneral(rows)
+                    global_rows_is = self.lg_map_fe.applyIS(rows_is)
+                    rows = global_rows_is.array
+                    I.setValues(rows, [idx], values)
+
+                I.assemble()  # lazy strategy for kron
+                interp_sub.append(I)
+            interp_1d.append(interp_sub)
+
+        # from IPython import embed; embed()
+        return interp_1d
+
+    def construct_kronecker_matrix(self, interp_1d):
```

Comment:
add comment saying it's based on the super().construct_kronecker_matrix but with modifications to ....